### ISSUE TITLE
[BC] deadlock when out-of-memory happens on some specific arrays, on top of pre-release branch

### ISSUE THREAD

**Issue by sgpyc** [gunrock/gunrock#201]
```
./bin/test_bc_8.0_x86_64 market /data/gunrock_dataset/large/soc-sinaweibo/soc-sinaweibo.mtx --src=randomize2 --in-sizing=1.1 --queue-sizing=1.2 --iteration-num=16 --64bit-SizeT --traversal-mode=LB --partition-method=biasrandom --device=0,1,2,3
Loading Matrix-market coordinate-formatted graph ...
  Reading directly from stored binary CSR arrays ...
Done reading (1s).

Degree Histogram (58655849 vertices, 522642066 edges):
    Degree   0: 0 (0.00%)
    Degree 2^0: 27041164 (46.10%)
    Degree 2^1: 16138616 (27.51%)
    Degree 2^2: 8472258 (14.44%)
    Degree 2^3: 4098677 (6.99%)
    Degree 2^4: 1760653 (3.00%)
    Degree 2^5: 646129 (1.10%)
    Degree 2^6: 237604 (0.41%)
    Degree 2^7: 91800 (0.16%)
    Degree 2^8: 63501 (0.11%)
    Degree 2^9: 47149 (0.08%)
    Degree 2^10: 23268 (0.04%)
    Degree 2^11: 12464 (0.02%)
    Degree 2^12: 19197 (0.03%)
    Degree 2^13: 2900 (0.00%)
    Degree 2^14: 380 (0.00%)
    Degree 2^15: 67 (0.00%)
    Degree 2^16: 21 (0.00%)
    Degree 2^17: 0 (0.00%)
    Degree 2^18: 1 (0.00%)

Using random source vertex for each run
Source vertex: 0
Using 4 GPU(s): [ 0 1 2 3 ].
partition end. (38601.267815 ms)
src_seed = 1475627195
Using traversal-mode LB
__________________________
3	 0	 0	 queue3 	 oversize :	 17298822 ->	 58655851
3	 1	 0	 queue3 	 oversize :	 17298822 ->	 58655851
2	 1	 0	 queue3 	 oversize :	 17297152 ->	 58655851
0	 1	 0	 queue3 	 oversize :	 18429916 ->	 58655851
2	 2	 0	 queue3 	 oversize :	 17297152 ->	 58655851
0	 2	 0	 queue3 	 oversize :	 18429916 ->	 58655851
1	 2	 0	 queue3 	 oversize :	 17361135 ->	 58655851
[../../gunrock/util/array_utils.cuh, 209 @ gpu 0] t_array cudaMalloc failed (CUDA error 2: out of memory)
(CUDA error 2 @ GPU 0: out of memory
(CUDA error 2 @ GPU 0: out of memory
(CUDA error 2 @ GPU 0: out of memory
(CUDA error 2 @ GPU 0: out of memory
(CUDA error 2 @ GPU 0: out of memory
(CUDA error 2 @ GPU 0: out of memory
(CUDA error 2 @ GPU 0: out of memory
1	 3	 0	 keys_in 	 oversize :	 6099959 ->	 14467611
2	 3	 0	 keys_in 	 oversize :	 5872300 ->	 14414292
3	 4	 0	 keys_in 	 oversize :	 5941380 ->	 14415684
0	 2	 1	 keys_in 	 oversize :	 7327318 ->	 15358262
1	 3	 0	 vertex_associate_in 	 oversize :	 6099959 ->	 14467611
2	 3	 0	 vertex_associate_in 	 oversize :	 5872300 ->	 14414292
3	 4	 0	 vertex_associate_in 	 oversize :	 5941380 ->	 14415684
0	 2	 1	 vertex_associate_in 	 oversize :	 7327318 ->	 15358262
[../../gunrock/util/array_utils.cuh, 209 @ gpu 0] vertex_associate_in[][] cudaMalloc failed (CUDA error 2: out of memory)
1	 3	 2	 keys_in 	 oversize :	 5758665 ->	 14467611
[../../gunrock/util/array_utils.cuh, 209 @ gpu 0] vertex_associate_in[][] cudaMalloc failed (CUDA error 2: out of memory)
3	 4	 1	 keys_in 	 oversize :	 5654069 ->	 14415684
2	 3	 0	 value__associate_in 	 oversize :	 11744600 ->	 14414292
0	 2	 1	 value__associate_in 	 oversize :	 14654636 ->	 15358262
1	 3	 2	 vertex_associate_in 	 oversize :	 5758665 ->	 14467611
3	 4	 1	 vertex_associate_in 	 oversize :	 5654069 ->	 14415684
2	 3	 1	 keys_in 	 oversize :	 5608036 ->	 14414292
1	 3	 2	 value__associate_in 	 oversize :	 11517330 ->	 14467611
0	 2	 2	 keys_in 	 oversize :	 7275396 ->	 15358262
3	 4	 1	 value__associate_in 	 oversize :	 11308138 ->	 14415684
2	 3	 1	 vertex_associate_in 	 oversize :	 5608036 ->	 14414292
0	 2	 2	 vertex_associate_in 	 oversize :	 7275396 ->	 15358262
[../../gunrock/util/array_utils.cuh, 209 @ gpu 2] vertex_associate_in[][] cudaMalloc failed (CUDA error 2: out of memory)
0	 2	 3	 keys_in 	 oversize :	 7301268 ->	 15358262
1	 3	 3	 keys_in 	 oversize :	 5779266 ->	 14467611
2	 3	 1	 value__associate_in 	 oversize :	 11216072 ->	 14414292
0	 2	 3	 vertex_associate_in 	 oversize :	 7301268 ->	 15358262
1	 3	 3	 vertex_associate_in 	 oversize :	 5779266 ->	 14467611
3	 4	 2	 keys_in 	 oversize :	 5586546 ->	 14415684
0	 2	 3	 value__associate_in 	 oversize :	 14602536 ->	 15358262
[../../gunrock/util/array_utils.cuh, 209 @ gpu 3] vertex_associate_in[][] cudaMalloc failed (CUDA error 2: out of memory)
2	 3	 3	 keys_in 	 oversize :	 5557749 ->	 14414292
3	 4	 2	 vertex_associate_in 	 oversize :	 5586546 ->	 14415684
[../../gunrock/util/array_utils.cuh, 209 @ gpu 2] vertex_associate_in[][] cudaMalloc failed (CUDA error 2: out of memory)
(CUDA error 2 @ GPU 12: out of memory
2	 3	 3	 vertex_associate_in 	 oversize :	 5557749 ->	 14414292
[../../gunrock/app/bc/bc_enactor.cuh, 1631 @ gpu 2] cudaStramWaitEvent failed (CUDA error 77: an illegal memory access was encountered)
[../../gunrock/app/bc/bc_enactor.cuh, 1631 @ gpu 2] cudaStramWaitEvent failed (CUDA error 77: an illegal memory access was encountered)
(CUDA error 77 @ GPU 12: an illegal memory access was encountered
[../../gunrock/app/bc/bc_enactor.cuh, 1631 @ gpu 0] cudaStramWaitEvent failed (CUDA error 77: an illegal memory access was encountered)
(CUDA error 77 @ GPU 12: an illegal memory access was encountered
(CUDA error 77 @ GPU 12: an illegal memory access was encountered
[../../gunrock/util/array_utils.cuh, 271 @ gpu 3] vertex_associate_in[][] cudaFree failed (CUDA error 77: an illegal memory access was encountered)
[../../gunrock/app/bc/bc_enactor.cuh, 1598 @ gpu 3] cudaEventRecord failed (CUDA error 77: an illegal memory access was encountered)
[../../gunrock/app/bc/bc_enactor.cuh, 1631 @ gpu 3] cudaStramWaitEvent failed (CUDA error 77: an illegal memory access was encountered)
[../../gunrock/app/bc/bc_enactor.cuh, 1631 @ gpu 3] cudaStramWaitEvent failed (CUDA error 77: an illegal memory access was encountered)
[../../gunrock/app/bc/bc_enactor.cuh, 1631 @ gpu 3] cudaStramWaitEvent failed (CUDA error 77: an illegal memory access was encountered)
```

**Comment by sgpyc** [comment]
Should be solved by ac2cb4a.
```
./bin/test_bc_8.0_x86_64 market /data/gunrock_dataset/large/soc-sinaweibo/soc-sinaweibo.mtx --src=randomize2 --in-sizing=1.1 --queue-sizing=1.2 --iteration-num=16 --64bit-SizeT --traversal-mode=LB --partition-method=biasrandom --device=0,1,2,3
Loading Matrix-market coordinate-formatted graph ...
  Reading directly from stored binary CSR arrays ...
Done reading (0s).

Degree Histogram (58655849 vertices, 522642066 edges):
    Degree   0: 0 (0.00%)
    Degree 2^0: 27041164 (46.10%)
    Degree 2^1: 16138616 (27.51%)
    Degree 2^2: 8472258 (14.44%)
    Degree 2^3: 4098677 (6.99%)
    Degree 2^4: 1760653 (3.00%)
    Degree 2^5: 646129 (1.10%)
    Degree 2^6: 237604 (0.41%)
    Degree 2^7: 91800 (0.16%)
    Degree 2^8: 63501 (0.11%)
    Degree 2^9: 47149 (0.08%)
    Degree 2^10: 23268 (0.04%)
    Degree 2^11: 12464 (0.02%)
    Degree 2^12: 19197 (0.03%)
    Degree 2^13: 2900 (0.00%)
    Degree 2^14: 380 (0.00%)
    Degree 2^15: 67 (0.00%)
    Degree 2^16: 21 (0.00%)
    Degree 2^17: 0 (0.00%)
    Degree 2^18: 1 (0.00%)

Using random source vertex for each run
Source vertex: 0
Using 4 GPU(s): [ 0 1 2 3 ].
partition end. (39618.791103 ms)
src_seed = 1475629927
Using traversal-mode LB
__________________________
0	 0	 0	 queue3 	 oversize :	 17442369 ->	 58655851
0	 1	 0	 queue3 	 oversize :	 17442369 ->	 58655851
1	 2	 0	 queue3 	 oversize :	 17399994 ->	 58655851
2	 2	 0	 queue3 	 oversize :	 18209048 ->	 58655851
3	 2	 0	 queue3 	 oversize :	 17335614 ->	 58655851
2	 3	 0	 queue3 	 oversize :	 18209048 ->	 58655851
3	 3	 0	 queue3 	 oversize :	 17335614 ->	 58655851
1	 3	 0	 queue3 	 oversize :	 17399994 ->	 58655851
[../../gunrock/util/array_utils.cuh, 209 @ gpu 2] t_array cudaMalloc failed (CUDA error 2: out of memory)
(CUDA error 2 @ GPU 0: out of memory
(CUDA error 2 @ GPU 0: out of memory
(CUDA error 2 @ GPU 0: out of memory
(CUDA error 2 @ GPU 0: out of memory
(CUDA error 2 @ GPU 0: out of memory
(CUDA error 2 @ GPU 0: out of memory
[test_bc.cu, 595 @ gpu 3] BC Problem Enact Failed (CUDA error 2: out of memory)

 [BC] finished.
 avg. elapsed: 0.0000 ms
 iterations: 0
 load time: 3114.6379 ms
 preprocess time: 47669.0202 ms
 postprocess time: 0.0000 ms
 total time: 51041.9650 ms
```

**Comment by yzhwang** [comment]
confirmed that the deadlock has gone.
